### ISSUE TITLE
Make Move framework test initialization functions idempotent

### ISSUE THREAD

**Issue by banool** [aptos-labs/aptos-core#8357]
So I’ve got the following code:
```
    #[test_only]
    fun create_test_account(
        mint_cap: &MintCapability<AptosCoin>,
        account: &signer,
    ) {
        account::create_account_for_test(signer::address_of(account));
        coin::register<AptosCoin>(account);
        let coins = coin::mint<AptosCoin>(100000, mint_cap);
        coin::deposit(signer::address_of(account), coins);
    }

    #[test_only]
    fun get_mint_cap(
        aptos_framework: &signer
    ): MintCapability<AptosCoin> {
        let (burn_cap, freeze_cap, mint_cap) = coin::initialize<AptosCoin>(
            aptos_framework,
            string::utf8(b"TC"),
            string::utf8(b"TC"),
            8,
            false,
        );
        coin::destroy_freeze_cap(freeze_cap);
        coin::destroy_burn_cap(burn_cap);
        mint_cap
    }

    #[test_only]
    fun create_thing(creator: &signer, friend1: &signer, friend2: &signer, aptos_framework: &signer): Object<Thing> {

        let mint_cap = get_mint_cap(aptos_framework);
        create_test_account(&mint_cap, creator);
        create_test_account(&mint_cap, friend1);
        create_test_account(&mint_cap, friend2);
        coin::destroy_mint_cap(mint_cap);

        create()
    }
```

As you can see, I use these get_mint_cap and create_test_account functions to create test accounts with funds.

This was working quite nicely up until now, when I’m trying to initialize testing stuff from another module as well in a test, specifically with functions from stake.move. Now I get the following error:
```
│ 435 │       fun initialize_internal<CoinType>(
│     │           ------------------- In this function in 0x1::coin
│     ·
│ 450 │ ╭         assert!(
│ 451 │ │             !exists<CoinInfo<CoinType>>(account_addr),
│ 452 │ │             error::already_exists(ECOIN_INFO_ALREADY_PUBLISHED),
│ 453 │ │         );
│     │ ╰─────────^ Test was not expected to error, but it aborted with code 524290 originating in the module 0000000000000000000000000000000000000000000000000000000000000001::coin rooted here
```

It seems like the cause is my code is trying to call coin::initialize but another module has already done it. I’ve tried initialize_for_test and initialize_for_test_without_aggregator_factory but those don’t fix it either.

The real fix is to have a function like this:
```
    #[test_only]
    fun create_test_account(
        aptos_framework: &signer,
        account: &signer,
    ) {
        if (!aptos_coin::has_mint_capability(aptos_framework)) {
            // If aptos_framework doesn't have the mint cap we need to initialize it.
            // This will initialize AptosCoin and store the mint cap in aptos_framwork.
            // These capabilities we have are just copies. We don't need them since we
            // use aptos_coin::mint to mint coins, which uses the mint cap from the
            // MintCapStore on aptos_framework. So we burn them.
            let (burn_cap, mint_cap) = aptos_coin::initialize_for_test(aptos_framework);
            coin::destroy_burn_cap(burn_cap);
            coin::destroy_mint_cap(mint_cap);
        };
        account::create_account_for_test(signer::address_of(account));
        coin::register<AptosCoin>(account);
        aptos_coin::mint(aptos_framework, signer::address_of(account), 100000);
    }
```

This is how the default function should work, rather than making the user deal with this.

**Comment by wrwg** [comment]
@banool this seems to be a complex feature, because there is really nothing special about test initializers. Do you have a suggestion how to implement this? Any example of another framework doing this idempotence?

**Comment by brmataptos** [comment]
A lot of test frameworks run each test in an isolated environment of some sort.  Simple mechanisms involve a lot of extra effort building and tearing down VMs.

**Comment by brmataptos** [comment]
Perhaps we could use a unique set of addresses for each test case, to provide isolation between them.  It would use more resources during testing to have multiple instances of each library module, but would probably be a lot faster than tearing down and rebuilding VMs for each test, or setting up parallel VMs for tests in parallel.
